LSP: Add skeleton support for Semantic Tokens

jhckragh opened the pull request:
This is a first step towards making the language server capable of providing useful semantic tokens.

There is probably room for improvement in the code as I'm not very familiar with Scala nor with the Flix compiler API.

The image below shows that the code is able to provide semantic tokens to the VS Code extension (once it has been extended as described [here](https://github.com/flix/vscode-flix/issues/122)).

![flix-st-example-1](https://user-images.githubusercontent.com/129259/133000581-8237a143-6ac4-44a5-927c-70ce36ea4726.png)

magnus-madsen commented:
Great work. I have some structural comments about the code, but otherwise the approach looks right to me!

magnus-madsen commented on main/src/ca/uwaterloo/flix/api/lsp/Index.scala:
```diff
@@ -178,6 +178,12 @@ case class Index(m: Map[(String, Int), List[Entity]],
+   * Returns all entities in the document at the given `uri`.
+   */
+  def query(uri: String): List[Entity] =
+    m.view.filterKeys { case (s, _) => s == uri }.values.flatten.toList
```
This is probably not very efficient with the use of `flatten` and `toList`. I guess it totally depends on: How many entities will there be? Maybe its not a practical problem.

magnus-madsen commented on main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala:
```diff
@@ -0,0 +1,99 @@
+   * sequence of integers, and the integer representing a token type is the type's
+   * index in the legend.
+   */
+  private object TokenType extends Enumeration {
```
I would represent this in a separate file like other enums, e.g.: https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/api/lsp/DocumentHighlightKind.scala

magnus-madsen commented on main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala:
```diff
@@ -0,0 +1,99 @@
+   * constraint applies to modifiers: The TS server must list the modifiers
+   * in the same order as is used here in the Scala code.
+   */
+  private object TokenModifier extends Enumeration {
```
Same comment- separate file and like https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/api/lsp/DocumentHighlightKind.scala

magnus-madsen commented on main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala:
```diff
@@ -0,0 +1,99 @@
+    val Declaration, Definition = Value
+  }
+
+  private case class SemanticToken(loc: SourceLocation,
```
Same comment.

magnus-madsen commented on main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala:
```diff
@@ -0,0 +1,99 @@
+package ca.uwaterloo.flix.api.lsp.provider
```
Add copyright header :)

magnus-madsen commented on main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala:
```diff
@@ -0,0 +1,99 @@
+                                   tokenType: SemanticTokensProvider.TokenType.TokenType,
+                                   tokenModifiers: List[SemanticTokensProvider.TokenModifier.TokenModifier])
+
+  def provideSemanticTokens(uri: String)(implicit index: Index, root: Root): JObject = {
```
Every function should have at least a one-line documentation comment :)

magnus-madsen commented on main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala:
```diff
@@ -0,0 +1,99 @@
+  }
+
+  private def getSemanticTokens(entity: Entity): List[SemanticToken] = entity match {
+    case Entity.Exp(e) => e match {
```
This approach looks right to me. I think you should be able to color everything but keywords (because they are not in the AST).

Do you know if SemanticTokens are additive? Are they layered on top of the grammer or do they replace it?

magnus-madsen replied:
It seems they are additive:

https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide

```
Semantic tokenization allows language servers to provide additional token information based on the language server's knowledge on how to resolve symbols in the context of a project. Themes can opt in to use semantic tokens to improve and refine the syntax highlighting from grammars.
```

I think this means that over time we want to have a simple grammar that provides basic syntax coloring (e.g. it could just be keywords) and then we can use semantic tokens to fill in the rest! :)

jhckragh replied:
From my experiments it seems that semantic highlighting takes precedence over highlighting provided by the `flix.tmLanguage.json` grammar file. For example, if the grammar file classifies a token as a variable and the language server classifies the token as a function, then the token will be highlighted as a function.

Since the language server knows the semantics of the file, it can provide more accurate highlighting than the simple grammar file. On the other hand, highlighting via the grammar file is faster than doing a round trip to the LS. So there's a tradeoff between speed and accuracy.

Also, from my experiments the LS classifies e.g. `Array.toList` as a function name (`Def` in compiler lingo?), but (to my mind at least) only `toList` should be classified as a function name, and `Array` should be classified as a namespace (like we do currently with the grammar file).

I think it's a nice idea to let semantic tokens drive the syntax highlighting. But probably some trial and error is necessary to get it right (and fast). 🙂

magnus-madsen replied:
OK, this supports my intuition that we should have a base grammar file and then leave more complicated stuff to semantic tokens.

magnus-madsen commented:
Once these structural issues are resolved, I am happy to merge this in as a "work in progress".

I would also, a bit later in the process, be interested to know what "tokens" we might be missing in "Index".

jhckragh commented:
Thank you for the kind words and the swift and thorough feedback! I will try to address your comments when I get home from work today. 🙂

jhckragh commented:
I've now made an attempt at addressing the review comments.

I've switched to an imperative implementation of the `query` method in order to avoid multiple traversals.

The tokens need to be sorted when they're encoded, and I'm not sure if I've expressed that part quite idiomatically.

Also, I pressed the "update branch" button in GitHub, and it did a merge instead of a rebase - sorry about that! 🙁 I can make a new PR with a clean history if required. (I don't know if it can be resolved in any another way - my knowledge of git is too limited, and I would probably just make things worse by trying various fixes.)

magnus-madsen commented:
> Also, I pressed the "update branch" button in GitHub, and it did a merge instead of a rebase - sorry about that! 🙁 I can make a new PR with a clean history if required. (I don't know if it can be resolved in any another way - my knowledge of git is too limited, and I would probably just make things worse by trying various fixes.)

I squash commits anyway, so thats usually not a problem.

magnus-madsen commented on main/src/ca/uwaterloo/flix/api/lsp/SemanticTokenType.scala:
```diff
@@ -0,0 +1,31 @@
+ * sequence of integers, and the integer representing a token type is the type's
+ * index in the legend.
+ */
+object SemanticTokenType extends Enumeration {
```
The use of `Enumeration` in Scala is actually deprecated. The recommended pattern is something like:

```
/**
  * Represents a `DiagnosticTag` in LSP.
  */
sealed trait DiagnosticTag {
  def toInt: Int = this match {
    case DiagnosticTag.Unnecessary => 1
    case DiagnosticTag.Deprecated => 2
  }
}

object DiagnosticTag {

  /**
    * Unused or unnecessary code.
    *
    * Clients are allowed to render diagnostics with this tag faded out instead of having an error squiggle.
    */
  case object Unnecessary extends DiagnosticTag

  /**
    * Deprecated or obsolete code.
    *
    * Clients are allowed to rendered diagnostics with this tag strike through.
    */
  case object Deprecated extends DiagnosticTag

}
```

magnus-madsen commented:
This looks good. I will merge once the tests complete!

jhckragh commented:
Thanks for your patient guidance! ☺️
